waiter/container_delete.go

```diff
@@ -0,0 +1,73 @@
+package waiter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/nspcc-dev/neofs-sdk-go/client"
+	apistatus "github.com/nspcc-dev/neofs-sdk-go/client/status"
+	"github.com/nspcc-dev/neofs-sdk-go/container"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+)
+
+// ContainerDeleteExecutor represents requirements to async container delete operation.
+type ContainerDeleteExecutor interface {
+	ContainerDelete(ctx context.Context, id cid.ID, prm client.PrmContainerDelete) error
+	ContainerGet(ctx context.Context, id cid.ID, prm client.PrmContainerGet) (container.Container, error)
+}
+
+// ContainerDeleteWaiter implements sync logic to container delete operation.
+type ContainerDeleteWaiter struct {
+	executor     ContainerDeleteExecutor
+	pollInterval time.Duration
+}
+
+// NewContainerDeleteWaiter is a constructor for ContainerDeleteWaiter.
+func NewContainerDeleteWaiter(executor ContainerDeleteExecutor) ContainerDeleteWaiter {
+	return ContainerDeleteWaiter{executor: executor, pollInterval: defaultPollInterval}
+}
+
+// SetPollInterval allows rewrite default poll interval.
+func (w *ContainerDeleteWaiter) SetPollInterval(interval time.Duration) {
+	w.pollInterval = interval
+}
+
+// ContainerDelete sends request to remove the NeoFS container.
+//
+// See details in [client.Client.ContainerDelete].
+func (w ContainerDeleteWaiter) ContainerDelete(ctx context.Context, id cid.ID, prm client.PrmContainerDelete) error {
+	if err := w.executor.ContainerDelete(ctx, id, prm); err != nil {
+		return nil
+	}
+
+	if err := w.isBucketDeleted(ctx, id); err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+func (w ContainerDeleteWaiter) isBucketDeleted(ctx context.Context, id cid.ID) error {
+	t := time.NewTicker(w.pollInterval)
+	defer t.Stop()
+
+	var cmdGet client.PrmContainerGet
+
+	for {
+		select {
+		case <-t.C:
+			_, err := w.executor.ContainerGet(ctx, id, cmdGet)
+			if err != nil {
+				if errors.Is(err, apistatus.ErrContainerNotFound) {
+					return nil
+				}
+
+				return fmt.Errorf("ContainerGet: %w", err)
+			}
```

Comment:
do we need to reset the ticker here? if an interval is `time.Second` and a request took 0.999 seconds i would expect it to wait for 1 second more rather than repeat the request immediately

Reply:
I tend to think, if request took 0.999s it is a system problem. For user important to understand - was the container deleted during this 0.999s or not. It doesn't matter if request took 0.999s or 0.001s, for user it is equal to `time.Second`

Reply:
> was the container deleted during this 0.999s or not

yes, but if not, then we send the request immediately and do not wait for 1 sec. in other words, i expect such `for` loops to work like "i will be requesting until it is ok but the time b/w the requests will be not less than 1 second". see how it [is done](https://github.com/nspcc-dev/neofs-node/blob/93733ff8905dfe8dac51a384007c7f7f73263e1a/cmd/neofs-cli/modules/container/create.go#L143) in `neof-cli`

@roman-khimov, @cthulhu-rider

Reply:
What is the "time between requests"? Suppose your RTT in 1s. You'll be polling once per two seconds instead of once per second this way. It can be handled both ways and practically you won't notice a difference in most of the scenarios, so I'd just stick to a simpler one.

Reply:
> What is the "time between requests"?

time b/w two sequential (the same) requests were initiated. ok, calling any `client.Client` method in other words

> Suppose your RTT in 1s.

i would say that smth bad is happening (and somebody is experiencing some load) and i would not call it as often as possible but would keep 1s idle b/w requests instead. also, i would consider the last response as an actual state so the next one can be requested with 1-sec delay

> I'd just stick to a simpler one

dont mind

Reply:
> time b/w two sequential (the same) requests were initiated

Then ticker works better already. What you're proposing is time between the last reply and a new request.